docs: document workflows (#2732)
SwissDataScienceCenter/renku: docs/topic-guides/workflows.rst

.. _workflows:

Renku Workflows
===============

> Together with :ref:`provenance of results <provenance>`, Renku workflows is
> a key feature of Renku to make code and processing pipelines reusable.
> When tracking executions with Renku, each execution creates a workflow step.
> These steps can be joined together into bigger workflows, making it easy to
> run them with varying inputs and on different execution backends.
> 
> There are many uses for this. You could define and test a pipeline locally
> and then execute it on an HPC cluster on a big dataset once you're happy.
> Or you could run your workflow offer a range of different parameters to see
> how the results differ. Or you create a workflow that runs daily with a
> CRON job.
> 
> All executions of workflows are also tracked in the provenance, without you
> having to re-type and memorize the commands used in each step.

Reviewer: ```suggestion
Workflows show how code and data in a project are connected. By encoding
these relationships, your project is easier for you to manage, and faster for
others to read and reuse!

Each time to track an execution with Renku, you create a workflow step.
Encoding a workflow step makes it easier for you to rerun it without retyping
long commands. Recording workflow steps in Renku also records
:ref:`provenance of results <provenance>`, metadata that you and others
can use to understand how a data file was generated.

To take full advantage of workflows, join individual steps together into multi-step
workflows. When your code pipeline is encoded as a workflow, you can
easily re-run all or portions of your workflow with simple commands, test
your code with different parameters and compare the results, and send it
to different execution backends. 

In addition, structuring your project code as a workflow makes it easy for
someone else to understand and get up to speed with your project. No
more reading through multiple files to understand how they are connected -
workflows make the connections between code and data files easy to
understand by listing each workflow step and its inputs and outputs. 
```

Reviewer 2: I forgot to merge this in initially but now added it separately.

Working with Workflows
----------------------

> Tracking execution of a command with :meth:`renku run <renku.ui.cli.run>`:

Reviewer: ```suggestion
To track your code execution as a Renku workflow, simply prepend :meth:`renku run <renku.ui.cli.run>` in front of your command. You may also identify the workflow step's inputs and outputs via the ``-i`` and ``-o`` flags, as shown here:
```

.. code-block:: console

>     $ renku run --name run-analysis -- python run_analysis.py -i inputs -o outputs

Reviewer: ```suggestion
    $ renku run --name run-analysis -- python run_analysis.py -i input_file.csv -o output_file.csv
```

Reviewer: I think the example is more concrete and intuitive if you show that the input and outputs are file paths

> creates a workflow step called `run-analysis`. You can inspect the workflow

Reviewer 2: do you mean to be using single ticks here (italics) instead of double ticks (code styling) ?

> creates a workflow step called `run-analysis`. You can inspect the workflow

Reviewer: ```suggestion
This command creates a workflow step called `run-analysis`. You can inspect the workflow
```

with :meth:`renku workflow show <renku.ui.cli.workflow>`:

.. code-block:: console

    $ renku workflow show run-analysis
    Id: /plans/76d73efb94964e9aac3635176ea57a36
    Name: run-analysis
    Creators: John Doe <example@renku.ch>

>     Command: python run_analysis.py -i inputs -o outputs
>     Success Codes:
>     Inputs:
>             - input-1:
>                     Default Value: run_analysis.py
>                     Position: 1
>             - i-2:
>                     Default Value: inputs
>                     Position: 2
>                     Prefix: -i
>     Outputs:
>             - o-3:
>                     Default Value: outputs

Reviewer: ```suggestion
    Command: python run_analysis.py -I input_file.csv -o output_file.csv
    Success Codes:
    Inputs:
            - input-1:
                    Default Value: run_analysis.py
                    Position: 1
            - i-2:
                    Default Value: input_file.csv
                    Position: 2
                    Prefix: -i
    Outputs:
            - o-3:
                    Default Value: output_file.csv
```

                    Position: 3
                    Prefix: -o

> You can execute it with :meth:`renku workflow execute <renku.ui.cli.workflow>`:

Reviewer: I recommend demonstrating basic re-execution of the workflow before introducing modifying the workflow parameters.

.. code-block:: console

    $ renku workflow execute run-analysis --set i-2=other_inputs

> which would run it on the file `other_inputs` instead of the original `inputs`

Reviewer 2: ```suggestion
which would run it on the file ``other_inputs`` instead of the original ``inputs``
```

file.

> You could also specify an execution backend with `--provider`, e.g. `toil` for
> execution in an HPC cluster (You need to install `renku` with the `toil` extra
> for this to be available).

Reviewer 2: ```suggestion
You could also specify an execution backend with ``--provider``, e.g. ``toil`` for
execution in an HPC cluster (You need to install ``renku`` with the ``toil`` extra
for this to be available).
```

> To create a workflow `my-workflow` out of multiple steps use :meth:`renku workflow compose <renku.ui.cli.workflow>`:

Reviewer 2: ```suggestion
To create a workflow ``my-workflow`` out of multiple steps use :meth:`renku workflow compose <renku.ui.cli.workflow>`:
```

> To create a workflow `my-workflow` out of multiple steps use :meth:`renku workflow compose <renku.ui.cli.workflow>`:

Reviewer 2: done
